crates/typst/src/geom/transform.rs

```diff
@@ -68,6 +68,68 @@ impl Transform {
     pub fn post_concat(self, next: Self) -> Self {
         next.pre_concat(self)
     }
+
+    /// Inverts the transformation.
+    ///
+    /// # Panics
+    /// If the determinant is zero.
+    pub fn invert(self) -> Self {
+        // Allow the trivial case to be inlined.
+        if self.is_identity() {
+            return self;
+        }
+
+        // Fast path for scale-translate-only transforms.
+        if self.kx.is_zero() && self.ky.is_zero() {
+            if self.sx.is_zero() || self.sy.is_zero() {
+                return Self::translate(-self.tx, -self.ty);
+            }
+
+            let inv_x = 1.0 / self.sx;
+            let inv_y = 1.0 / self.sy;
+            return Self {
+                sx: Ratio::new(inv_x),
+                ky: Ratio::zero(),
+                kx: Ratio::zero(),
+                sy: Ratio::new(inv_y),
+                tx: -self.tx * inv_x,
+                ty: -self.ty * inv_y,
+            };
+        }
+
+        let det = self.sx * self.sy - self.kx * self.ky;
+        if det.get() < 1e-12 {
+            panic!("Transform is not invertible");
+        }
```

Comment:
Is it truly impossible for transformations in Typst to be irreversible?

Reply:
I’ll double check, but I think not, it would require the scale to be zero, something I could just make an error into `transform` and `scale` functions, basically preventing the condition earlier in the chain (since it’s required for PNG rendering too and it’s only in cases that are basically nonsensical for typst anyways.

Reply:
This should return `Option<Self>` in my opinion. If at all, it's better to unwrap at a call site that has more context information.